x-pack/agent/dev-tools/cmd/fakewebapi/main.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/x-pack/agent/pkg/fleetapi"
+)
+
+var (
+	host   string
+	apiKey string
+	mutex  sync.Mutex
+
+	pathCheckin     = regexp.MustCompile(`^/api/fleet/agents/(.+)/checkin`)
+	checkinResponse = response{Actions: make([]action, 0), Success: true}
+)
+
+type response struct {
+	Actions []action `json:"actions"`
+	Success bool     `json:"success"`
+}
+
+type action interface{}
+
+func init() {
+	flag.StringVar(&apiKey, "apikey", "abc123", "API Key to authenticate")
+	flag.StringVar(&host, "host", "localhost:8080", "The IP and port to use for the webserver")
+}
+
+func main() {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/api/fleet/agents/enroll", handlerEnroll)
+	mux.HandleFunc("/admin/actions", handlerAction)
+	mux.HandleFunc("/", handlerRoot)
+
+	log.Printf("Starting webserver and listening on %s", host)
+
+	listener, err := net.Listen("tcp", host)
+	if err != nil {
+		panic(err)
+	}
+	defer listener.Close()
+
+	http.Serve(listener, mux)
+}
+
+func handlerRoot(w http.ResponseWriter, r *http.Request) {
+	if pathCheckin.MatchString(r.URL.Path) {
+		authHandler(handlerCheckin, apiKey)(w, r)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(`{ "message": "Hello!"}`))
+	log.Println("Root hello!")
+	log.Println("Path: ", r.URL.Path)
+}
+
+func handlerEnroll(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, "Bad Request", http.StatusBadRequest)
+		return
+	}
+
+	response := &fleetapi.EnrollResponse{
+		Action:  "created",
+		Success: true,
+		Item: fleetapi.EnrollItemResponse{
+			ID:                   "a4937110-e53e-11e9-934f-47a8e38a522c",
+			Active:               true,
+			PolicyID:             "default",
+			Type:                 fleetapi.PermanentEnroll,
+			EnrolledAt:           time.Now(),
+			UserProvidedMetadata: make(map[string]interface{}),
+			LocalMetadata:        make(map[string]interface{}),
+			AccessAPIKey:         apiKey,
+		},
+	}
+
+	b, err := json.Marshal(response)
+	if err != nil {
+		log.Printf("failed to enroll: %+v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	w.Write(b)
+	log.Println("Enroll response:", string(b))
+}
+
+func handlerCheckin(w http.ResponseWriter, r *http.Request) {
+	mutex.Lock()
```

Comment:
what this lock locks?
you are responding with same response always, what about preparing mashalled response upfront

Reply:
Well, I wanted to make it more freeform, you can actually push a new configuration using the `handlerAction`